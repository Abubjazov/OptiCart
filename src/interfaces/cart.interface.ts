import { Product } from './product.interface'

export enum CartActionTypes {
	FETCH_CART = 'FETCH_CART',
	FETCH_CART_SUCCESS = 'FETCH_CART_SUCCESS',
	FETCH_CART_ERROR = 'FETCH_CART_ERROR',
	ADD_TO_CART = 'ADD_TO_CART',
	ADD_TO_CART_SUCCESS = 'ADD_TO_CART_SUCCESS',
	ADD_TO_CART_ERROR = 'ADD_TO_CART_ERROR',
	REMOVE_FROM_CART = 'REMOVE_FROM_CART',
	REMOVE_FROM_CART_SUCCESS = 'REMOVE_FROM_CART_SUCCESS',
	REMOVE_FROM_CART_ERROR = 'REMOVE_FROM_CART_ERROR',
	UPDATE_CART_QUANTITY = 'UPDATE_CART_QUANTITY',
	UPDATE_CART_QUANTITY_SUCCESS = 'UPDATE_CART_QUANTITY_SUCCESS',
	UPDATE_CART_QUANTITY_ERROR = 'UPDATE_CART_QUANTITY_ERROR',
}

export interface CartListItem extends Product {
	product_id: number
	quantity: number
}

export interface CartState {
	cart: CartListItem[]
	currentItemId: null | number
	status: 'waiting' | 'error' | 'loading'
	error: null | string
}

interface FetchCartAction {
	type: CartActionTypes.FETCH_CART
}

interface FetchCartSuccessAction {
	type: CartActionTypes.FETCH_CART_SUCCESS
	payload: CartListItem[]
}

interface FetchCartErrorAction {
	type: CartActionTypes.FETCH_CART_ERROR
	payload: string
}

interface addToCartAction {
	type: CartActionTypes.ADD_TO_CART
	payload: number
}

interface addToCartSuccessAction {
	type: CartActionTypes.ADD_TO_CART_SUCCESS
	payload: CartListItem[]
}
interface addToCartErrorAction {
	type: CartActionTypes.ADD_TO_CART_ERROR
	payload: string
}
// interface Action {
// 	type: CartActionTypes
// 	payload: any
// }
// interface Action {
// 	type: CartActionTypes
// 	payload: any
// }
// interface Action {
// 	type: CartActionTypes
// 	payload: any
// }
// interface Action {
// 	type: CartActionTypes
// 	payload: any
// }
// interface Action {
// 	type: CartActionTypes
// 	payload: any
// }
// interface Action {
// 	type: CartActionTypes
// 	payload: any
// }

export type CartAction =
	| FetchCartAction
	| FetchCartSuccessAction
	| FetchCartErrorAction
	| addToCartAction
	| addToCartSuccessAction
	| addToCartErrorAction
