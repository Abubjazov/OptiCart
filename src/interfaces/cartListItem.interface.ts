import { Product } from './product.interface'

export enum CartActionTypes {
	FETCH_CART = 'FETCH_CART',
	FETCH_CART_SUCCESS = 'FETCH_CART_SUCCESS',
	FETCH_CART_ERROR = 'FETCH_CART_ERROR',
}

export interface CartListItem extends Product {
	product_id: number
	quantity: number
}

export interface CartState {
	cart: CartListItem[]
	status: 'waiting' | 'error' | 'loading'
	error: null | string
}

interface FetchCartAction {
	type: CartActionTypes.FETCH_CART
}

interface FetchCartSuccessAction {
	type: CartActionTypes.FETCH_CART_SUCCESS
	payload: CartListItem[]
}

interface FetchCartErrorAction {
	type: CartActionTypes.FETCH_CART_ERROR
	payload: string
}

export type CartAction =
	| FetchCartAction
	| FetchCartSuccessAction
	| FetchCartErrorAction
