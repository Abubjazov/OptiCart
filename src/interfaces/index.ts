export * from './cartListItem.interface'
export * from './product.interface'
export * from './action.interface'
