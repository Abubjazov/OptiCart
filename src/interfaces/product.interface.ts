export enum ProductActionTypes {
	FETCH_PRODUCTS = 'FETCH_PRODUCTS',
	FETCH_PRODUCTS_SUCCESS = 'FFETCH_PRODUCTS_SUCCESS',
	FETCH_PRODUCTS_ERROR = 'FETCH_PRODUCTS_ERROR',
}

export interface Product {
	id: number
	name: string
	picture: string
	picture_medium: string
	picture_small: string
	description: string
	price: number
}

export interface ProductState {
	products: Product[]
	currentProductId: null | number
	status: 'waiting' | 'error' | 'loading'
	error: null | string
}

interface FetchProductsAction {
	type: ProductActionTypes.FETCH_PRODUCTS
}

interface FetchProductsSuccessAction {
	type: ProductActionTypes.FETCH_PRODUCTS_SUCCESS
	payload: Product[]
}

interface FetchProductsErrorAction {
	type: ProductActionTypes.FETCH_PRODUCTS_ERROR
	payload: string
}

export type ProductAction =
	| FetchProductsAction
	| FetchProductsSuccessAction
	| FetchProductsErrorAction
