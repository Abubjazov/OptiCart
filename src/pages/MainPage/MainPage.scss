@import '../../variables';

.main-page {
	background: $gray;

	.container {
		padding: 11vh 0 1vh 0;
		min-height: 100vh;
		max-width: calc(100vw - 2vw);
		display: grid;
		grid-template-columns: repeat(auto-fit, minmax(210px, 1fr));
		row-gap: 10px;
		justify-items: center;
	}
}
