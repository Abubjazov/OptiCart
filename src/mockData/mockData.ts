import { CartListItem } from '../interfaces'

export const cartItems: CartListItem[] = [
	{
		id: 2,
		product_id: 15,
		name: 'Rangoto Vi',
		picture: 'https://i.ibb.co/T2XSnw0/z9.webp',
		description: 'Nemo dolore officia est natus.',
		price: 11.04,
		quantity: 3,
	},
	{
		id: 3,
		product_id: 16,
		name: 'Armo Sento 2',
		picture: 'https://i.ibb.co/TPT93hh/z10.jpg',
		description: 'Nobis in voluptatum qui voluptate.',
		price: 17.22,
		quantity: 1,
	},
	{
		id: 4,
		product_id: 9,
		name: 'Vertigo 2',
		picture: 'https://i.ibb.co/vvCZ6gp/nike.webp',
		description: 'Architecto omnis sed nostrum aliquid.',
		price: 26.11,
		quantity: 2,
	},
]
