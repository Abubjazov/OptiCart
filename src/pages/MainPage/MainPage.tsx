import { nanoid } from 'nanoid'

import { ProductCard } from '../../components'
import { Product } from '../../interfaces/product.interface'

import './MainPage.scss'

export const MainPage = (): JSX.Element => {
	const products: Product[] = [
		{
			id: 1,
			name: 'Air 1',
			picture: 'https://i.ibb.co/Y281B6p/nike9.webp',
			description: 'Qui animi et ullam dignissimos.',
			price: 12.32,
		},
		{
			id: 2,
			name: 'Air 2',
			picture: 'https://i.ibb.co/HFLmzLp/nike8.webp',
			description: 'Soluta ratione assumenda accusamus sequi.',
			price: 10.21,
		},
		{
			id: 3,
			name: 'Rango Max',
			picture: 'https://i.ibb.co/R9FwhBx/nike7.webp',
			description: 'Est odit minima ut sunt.',
			price: 9.15,
		},
		{
			id: 4,
			name: 'Paco',
			picture: 'https://i.ibb.co/zGd7cKQ/nike6.webp',
			description: 'Nulla nihil adipisci accusantium suscipit.',
			price: 16.14,
		},
		{
			id: 5,
			name: 'Indigo',
			picture: 'https://i.ibb.co/WKpMdP9/nike5.webp',
			description: 'Earum sit repellat eos rem.',
			price: 27.17,
		},
		{
			id: 6,
			name: 'Nimiro',
			picture: 'https://i.ibb.co/27LxqHH/nike4.webp',
			description: 'Dolor optio dicta rem quia.',
			price: 23.12,
		},
		{
			id: 7,
			name: 'RGenta',
			picture: 'https://i.ibb.co/YQcJ4D4/nike3.jpg',
			description: 'Exercitationem magnam illo distinctio enim.',
			price: 16.34,
		},
		{
			id: 8,
			name: 'Vertigo 1',
			picture: 'https://i.ibb.co/M7BdDB0/nike2.webp',
			description: 'Ea vel eum ullam cupiditate.',
			price: 12.54,
		},
		{
			id: 9,
			name: 'Vertigo 2',
			picture: 'https://i.ibb.co/vvCZ6gp/nike.webp',
			description: 'Unde fugit asperiores eum voluptas.',
			price: 26.11,
		},
		{
			id: 10,
			name: 'Orbito',
			picture: 'https://i.ibb.co/tK0hN6Q/nike11.webp',
			description: 'Numquam excepturi consequatur voluptates quisquam.',
			price: 36.67,
		},
	]

	return (
		<main className='main-page'>
			<div className='container'>
				{products.map((item: Product) => (
					<ProductCard key={nanoid()} {...item} />
				))}
			</div>
		</main>
	)
}
