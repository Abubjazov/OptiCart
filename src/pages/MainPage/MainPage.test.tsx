import { render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { Provider } from 'react-redux'
import configureStore from 'redux-mock-store'
import thunk from 'redux-thunk'

import { CartActionTypes, ProductActionTypes } from '../../interfaces'

import { mockProducts } from '../../mockData/mockData'

import { MainPage } from './MainPage'

const middlewares = [thunk]
const mockStore = configureStore(middlewares)

describe('Component: MainPage', () => {
	test('should render MainPage with Spinner', () => {
		const store = mockStore({
			product: {
				status: 'loading',
				currentProductId: null,
			},
			cart: {
				cart: [],
				status: 'waiting',
				currentItemId: null,
			},
		})

		const { asFragment } = render(
			<Provider store={store}>
				<MainPage />
			</Provider>
		)

		expect(asFragment()).toMatchSnapshot()
	})

	test('should render MainPage with Error', () => {
		const store = mockStore({
			product: {
				status: 'error',
				error: 'Test error message',
			},
			cart: {
				cart: [],
				status: 'waiting',
				currentItemId: null,
			},
		})

		const { asFragment } = render(
			<Provider store={store}>
				<MainPage />
			</Provider>
		)

		expect(asFragment()).toMatchSnapshot()
	})

	test('should render MainPage with products list', () => {
		const store = mockStore({
			product: {
				products: mockProducts,
				status: 'waiting',
				currentProductId: null,
			},
			cart: {
				cart: [],
				status: 'loading',
				currentItemId: 1,
			},
		})

		const { asFragment } = render(
			<Provider store={store}>
				<MainPage />
			</Provider>
		)

		expect(asFragment()).toMatchSnapshot()
	})

	test('should dispatch an action on "Add to cart" button click', async () => {
		const store = mockStore({
			product: {
				products: mockProducts,
				status: 'waiting',
				currentProductId: null,
				error: null,
			},
			cart: {
				cart: [],
				status: 'waiting',
				currentItemId: null,
			},
		})
		const expectedActions = [
			{ type: ProductActionTypes.FETCH_PRODUCTS },
			{ type: CartActionTypes.ADD_TO_CART, payload: 1 },
		]
		render(
			<Provider store={store}>
				<MainPage />
			</Provider>
		)
		userEvent.click(screen.getAllByText('Add to cart')[0])
		expect(store.getActions()).toEqual(expectedActions)
	})
})
