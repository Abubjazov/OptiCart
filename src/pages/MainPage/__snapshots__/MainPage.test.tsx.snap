// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: MainPage should render MainPage with Error 1`] = `
<DocumentFragment>
  <main>
    <div
      class="error-message"
      role="alert"
    >
      <p
        tabindex="0"
      >
        Test error message
      </p>
      <p
        tabindex="0"
      />
      <a
        href="/optiCart/"
      >
        Back to main page
      </a>
    </div>
  </main>
</DocumentFragment>
`;

exports[`Component: MainPage should render MainPage with Spinner 1`] = `
<DocumentFragment>
  <main
    class="main-page"
  >
    <div
      class="container"
    >
      <svg
        class="spinner"
        height="200"
        preserveAspectRatio="xMidYMid"
        viewBox="0 0 100 100"
        width="200"
      >
        <path
          d="M24.3 30C11.4 30 5 43.3 5 50s6.4 20 19.3 20c19.3 0 32.1-40 51.4-40 C88.6 30 95 43.3 95 50s-6.4 20-19.3 20C56.4 70 43.6 30 24.3 30z"
          fill="none"
          stroke="#3b8079"
          stroke-dasharray="42.76482137044271 42.76482137044271"
          stroke-linecap="round"
          stroke-width="8"
        >
          <animate
            attributeName="stroke-dashoffset"
            dur="1s"
            keyTimes="0;1"
            repeatCount="indefinite"
            values="0;256.58892822265625"
          />
        </path>
      </svg>
    </div>
  </main>
</DocumentFragment>
`;

exports[`Component: MainPage should render MainPage with products list 1`] = `
<DocumentFragment>
  <main
    class="main-page"
  >
    <p
      role="alert"
      style="display: block; position: fixed; left: 100px; top: 0px; overflow: hidden; height: 0px; opacity: 1;"
    >
      You are on the main page
    </p>
    <div
      class="container"
    >
      <article
        aria-label="sneakers Paco, unit price 54.5 $"
        class="product"
        tabindex="0"
      >
        <header>
          <img
            alt="Paco"
            height="200"
            src="https://i.ibb.co/84RrCYH/4-400-400.webp"
            width="200"
          />
        </header>
        <section
          class="product-descriptor"
        >
          <h2>
            Paco
          </h2>
          <p>
            Ab odit numquam consectetur aut.
          </p>
        </section>
        <footer>
          54.5 $
          <button
            aria-label="Add to cart Paco"
          >
            <svg
              class="small-spinner"
              height="20"
              preserveAspectRatio="xMidYMid"
              viewBox="0 0 100 100"
              width="20"
            >
              <path
                d="M24.3 30C11.4 30 5 43.3 5 50s6.4 20 19.3 20c19.3 0 32.1-40 51.4-40 C88.6 30 95 43.3 95 50s-6.4 20-19.3 20C56.4 70 43.6 30 24.3 30z"
                fill="none"
                stroke="#fff"
                stroke-dasharray="42.76482137044271 42.76482137044271"
                stroke-linecap="round"
                stroke-width="8"
              >
                <animate
                  attributeName="stroke-dashoffset"
                  dur="1s"
                  keyTimes="0;1"
                  repeatCount="indefinite"
                  values="0;256.58892822265625"
                />
              </path>
            </svg>
          </button>
        </footer>
      </article>
      <article
        aria-label="sneakers Indigo, unit price 16.63 $"
        class="product"
        tabindex="0"
      >
        <header>
          <img
            alt="Indigo"
            height="200"
            src="https://i.ibb.co/FBTQYHv/5-400-400.webp"
            width="200"
          />
        </header>
        <section
          class="product-descriptor"
        >
          <h2>
            Indigo
          </h2>
          <p>
            Rem ut excepturi quidem nesciunt.
          </p>
        </section>
        <footer>
          16.63 $
          <button
            aria-label="Add to cart Indigo"
          >
            Add to cart
          </button>
        </footer>
      </article>
      <article
        aria-label="sneakers Nimiro, unit price 22.85 $"
        class="product"
        tabindex="0"
      >
        <header>
          <img
            alt="Nimiro"
            height="200"
            src="https://i.ibb.co/wwjGRcp/6-400-400.webp"
            width="200"
          />
        </header>
        <section
          class="product-descriptor"
        >
          <h2>
            Nimiro
          </h2>
          <p>
            Quasi cupiditate dolorum nulla tempore.
          </p>
        </section>
        <footer>
          22.85 $
          <button
            aria-label="Add to cart Nimiro"
          >
            Add to cart
          </button>
        </footer>
      </article>
    </div>
  </main>
</DocumentFragment>
`;
